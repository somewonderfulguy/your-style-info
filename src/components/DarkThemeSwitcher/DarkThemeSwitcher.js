import React from 'react'

const DarkThemeSwitcher = () => (
  <div>
    Switch
  </div>
)

export default DarkThemeSwitcher