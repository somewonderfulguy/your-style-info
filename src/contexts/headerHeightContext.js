import React, {createContext, useContext, useMemo, useState} from 'react'

export const ERROR_HEADER_HEIGHT = 'useHeaderHeight must be used within a HeaderHeightProvider'

const HeaderHeightContext = createContext()

const useHeaderHeight = () => {
  const context = useContext(HeaderHeightContext)
  if(!context) {
    throw new Error(ERROR_HEADER_HEIGHT)
  }
  return context
}

const HeaderHeightProvider = props => {
  const [headerHeight, setHeaderHeight] = useState(0)
  const value = useMemo(() => ({headerHeight, setHeaderHeight}), [headerHeight])
  return <HeaderHeightContext.Provider value={value} {...props} />
}

export {HeaderHeightProvider, useHeaderHeight}