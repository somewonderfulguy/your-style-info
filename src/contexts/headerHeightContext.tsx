import {
  createContext,
  Dispatch,
  ReactNode,
  SetStateAction,
  useContext,
  useState
} from 'react'

const HeaderHeightStateContext = createContext<number | undefined>(undefined)
HeaderHeightStateContext.displayName = 'HeaderHeightStateContext'

const HeaderHeightDispatchContext = createContext<
  Dispatch<SetStateAction<number>> | undefined
>(undefined)
HeaderHeightDispatchContext.displayName = 'HeaderHeightDispatchContext'

const HeaderHeightProvider = ({ children }: { children: ReactNode }) => {
  const [headerHeight, setHeaderHeight] = useState(0)
  return (
    <HeaderHeightStateContext.Provider value={headerHeight}>
      <HeaderHeightDispatchContext.Provider value={setHeaderHeight}>
        {children}
      </HeaderHeightDispatchContext.Provider>
    </HeaderHeightStateContext.Provider>
  )
}

const useHeaderHeightState = () => {
  const context = useContext(HeaderHeightStateContext)
  if (context === undefined) {
    throw new Error(
      'useHeaderHeightState must be used within a HeaderHeightProvider'
    )
  }
  return context
}

const useHeaderHeightDispatch = () => {
  const context = useContext(HeaderHeightDispatchContext)
  if (context === undefined) {
    throw new Error(
      'useHeaderHeightDispatch must be used within a HeaderHeightProvider'
    )
  }
  return context
}

export { HeaderHeightProvider, useHeaderHeightState, useHeaderHeightDispatch }
