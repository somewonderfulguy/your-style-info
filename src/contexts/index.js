export * from './headerHeightContext'
export * from './loadingContext'
export * from './localisationContext'
export * from './screenDimensionsContext'
export * from './themeContext'