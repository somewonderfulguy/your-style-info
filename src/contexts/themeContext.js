import React, {createContext, useCallback, useContext, useMemo, useEffect, useState} from 'react'

const ThemeContext = createContext()

const useTheme = () => {
  const context = useContext(ThemeContext)
  if(!context) {
    throw new Error('useTheme must be used within a ThemeProvider')
  }
  const [isDarkTheme, setDarkTheme] = context

  const switchTheme = useCallback(
    () => setDarkTheme(!isDarkTheme),
    [isDarkTheme, setDarkTheme]
  )

  return {isDarkTheme, switchTheme}
}

const ThemeProvider = props => {
  const preferDarkQuery = '(prefers-color-scheme: dark)'
  const [isDarkTheme, setDarkTheme] = useState(
    () =>
      window.localStorage.getItem('isDarkTheme') === 'true' ||
      (window.matchMedia(preferDarkQuery).matches ? true : false),
  )

  useEffect(() => {
    const mediaQuery = window.matchMedia(preferDarkQuery)
    const handleChange = () => setDarkTheme(mediaQuery.matches ? true : false)
    mediaQuery.addListener(handleChange)
    return () => mediaQuery.removeListener(handleChange)
  }, [])

  useEffect(() => {
    window.localStorage.setItem('isDarkTheme', isDarkTheme)
  }, [isDarkTheme])

  const value = useMemo(() => [isDarkTheme, setDarkTheme], [isDarkTheme])

  return <ThemeContext.Provider value={value} {...props} />
}

export {ThemeProvider, useTheme}