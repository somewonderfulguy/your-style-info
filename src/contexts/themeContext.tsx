import React, {
  createContext, Dispatch, ReactNode, SetStateAction,
  useCallback, useContext, useEffect, useState
} from 'react'

type dispatchType = Dispatch<SetStateAction<boolean>>

const ThemeStateContext = createContext<boolean | undefined>(undefined)
ThemeStateContext.displayName = 'ThemeStateContext'

const ThemeDispatchContext = createContext<dispatchType | undefined>(undefined)
ThemeDispatchContext.displayName = 'ThemeDispatchContext'

const ThemeProvider = ({ children }: { children: ReactNode | ReactNode[] }) => {
  const preferDarkQuery = '(prefers-color-scheme: dark)'
  const [isDarkTheme, setDarkTheme] = useState(
    () =>
      window.localStorage.getItem('isDarkTheme') === 'true' ||
      (window.matchMedia(preferDarkQuery).matches ? true : false)
  )

  useEffect(() => {
    const mediaQuery = window.matchMedia(preferDarkQuery)
    const handleChange = () => setDarkTheme(mediaQuery.matches ? true : false)
    mediaQuery.addListener(handleChange)
    return () => mediaQuery.removeListener(handleChange)
  }, [])

  useEffect(() => {
    window.localStorage.setItem('isDarkTheme', String(isDarkTheme))
  }, [isDarkTheme])

  return (
    <ThemeStateContext.Provider value={isDarkTheme}>
      <ThemeDispatchContext.Provider value={setDarkTheme}>
        {children}
      </ThemeDispatchContext.Provider>
    </ThemeStateContext.Provider>
  )
}

const useThemeState = () => {
  const isDarkTheme = useContext(ThemeStateContext)

  if(isDarkTheme === undefined) {
    throw new Error('useThemeState must be used within a ThemeProvider')
  }

  return isDarkTheme
}

const useThemeSwitch = () => {
  const setDarkTheme = useContext(ThemeDispatchContext)

  if(setDarkTheme === undefined) {
    throw new Error('useThemeSwitch must be used within a ThemeProvider')
  }

  const switchTheme = useCallback(
    () => setDarkTheme(prevTheme => !prevTheme),
    [setDarkTheme]
  )

  return switchTheme
}

export {ThemeProvider, useThemeState, useThemeSwitch}