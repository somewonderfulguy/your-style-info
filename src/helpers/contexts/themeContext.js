import React, {createContext, useContext, useMemo, useEffect, useState} from 'react'

const ThemeContext = createContext()

const useTheme = () => {
  const context = useContext(ThemeContext)
  if(!context) {
    throw new Error('useTheme must be used within a ThemeProvider')
  }
  const [isDarkTheme, setDarkTheme] = context

  const switchTheme = () => {
    setDarkTheme(!isDarkTheme)
    localStorage.setItem('isDarkTheme', !isDarkTheme)
  }

  return {isDarkTheme, switchTheme}
}

const ThemeProvider = props => {
  const [isDarkTheme, setDarkTheme] = useState(false)
  const [hasThemeLoaded, setThemeLoaded] = useState(false)

  useEffect(() => {
    const isLocalStorageDark = localStorage.getItem('isDarkTheme')
    setDarkTheme(isLocalStorageDark === 'true')
    setThemeLoaded(true)
  }, [])

  const value = useMemo(() => [isDarkTheme, setDarkTheme], [isDarkTheme])

  if(!hasThemeLoaded) {
    // avoiding render default light theme and then re-render to
    // dark theme if localStorage keeps dark mode as true
    return <div />
  }

  return <ThemeContext.Provider value={value} {...props} />
}

export {ThemeProvider, useTheme}