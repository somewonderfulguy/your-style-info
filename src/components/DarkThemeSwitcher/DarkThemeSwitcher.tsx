import { useThemeState, useThemeSwitch } from '~contexts/themeContext'

// TODO: colocate images with component
import moonIcon from '~assets/images/moon.svg'
import sunIcon from '~assets/images/sun.svg'
import styles from './DarkThemeSwitcher.module.css'

type Props = {
  darkerPalette?: boolean
  labelText?: boolean
}

const DarkThemeSwitcher = ({
  darkerPalette = false,
  labelText = false
}: Props) => {
  const isDarkTheme = useThemeState()
  const switchTheme = useThemeSwitch()

  console.log(moonIcon)

  return (
    // TODO: dynamically change aria label text
    <label
      className={darkerPalette ? styles.switchDarker : styles.switch}
      aria-label="switch theme"
    >
      <input
        className={styles.checkbox}
        type="checkbox"
        checked={isDarkTheme}
        onChange={switchTheme}
      />
      <div className={styles.slider}>
        <div
          className={styles.round}
          style={{
            backgroundImage: `url("${sunIcon.src}")`,
            opacity: isDarkTheme ? 0 : 1
          }}
        />
        <div
          className={styles.round}
          style={{
            backgroundImage: `url("${moonIcon.src}")`,
            opacity: isDarkTheme ? 1 : 0,
            zIndex: 1
          }}
        />
      </div>
      {labelText && <span className={styles.text}>Switch theme</span>}
    </label>
  )
}

export default DarkThemeSwitcher
