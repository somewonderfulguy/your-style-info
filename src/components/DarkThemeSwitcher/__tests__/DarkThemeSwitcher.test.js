import React from 'react'
import {screen, render} from '@testing-library/react'
import {renderHook} from '@testing-library/react-hooks'
import user from '@testing-library/user-event'

import * as mockThemeContext from 'contexts/themeContext'
import {ThemeProvider, useTheme} from 'contexts'
import DarkThemeSwitcher from '..'

jest.spyOn(mockThemeContext, 'useTheme')
afterEach(() => jest.clearAllMocks())

const setup = (darkerPalette = false, labelText = false) => {
  const utils = render (
    <DarkThemeSwitcher darkerPalette={darkerPalette} labelText={labelText} />, {
      wrapper: ThemeProvider
    }
  )
  return utils
}

test('should match snapshot', () => {
  const {asFragment} = setup()
  expect(asFragment()).toMatchSnapshot()
})

test('should change depending on darkerPalette prop', () => {
  const defaultProps = setup().asFragment()
  const withDarkPalette = setup(true).asFragment()

  expect(defaultProps).toMatchDiffSnapshot(withDarkPalette, {
    contextLines: 1,
    aAnnotation: 'dark palette off',
    bAnnotation: 'dark palette on'
  })
})

test('should change depending on labelText prop', () => {
  const defaultProps = setup().asFragment()
  const withLabelText = setup(false, true).asFragment()

  expect(defaultProps).toMatchDiffSnapshot(withLabelText, {
    contextLines: 1,
    aAnnotation: 'label text off',
    bAnnotation: 'label text on'
  })
})

test('switching theme should work properly', async () => {
  const {asFragment} = setup()
  const getHook = () => renderHook(() => useTheme(), {
    wrapper: ThemeProvider
  })
  const checkbox = screen.getByRole('checkbox')

  // default (light)
  expect(getHook().result.current.isDarkTheme).toBeFalsy()
  expect(window.localStorage.getItem('isDarkTheme')).toEqual('false')

  // switch to dark
  user.click(checkbox)
  expect(getHook().result.current.isDarkTheme).toBeTruthy()
  expect(window.localStorage.getItem('isDarkTheme')).toEqual('true')
  const darkComponent = asFragment()

  // switch to light
  user.click(checkbox)
  expect(getHook().result.current.isDarkTheme).toBeFalsy()
  expect(window.localStorage.getItem('isDarkTheme')).toEqual('false')
  const lightComponent = asFragment()

  // snapshot difference
  expect(darkComponent).toMatchDiffSnapshot(lightComponent, {
    contextLines: 2,
    aAnnotation: 'dark theme on',
    bAnnotation: 'light theme on'
  })
})