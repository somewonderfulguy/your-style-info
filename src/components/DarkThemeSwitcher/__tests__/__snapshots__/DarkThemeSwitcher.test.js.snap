// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should change depending on darkerPalette prop 1`] = `
Snapshot Diff:
- dark palette off
+ dark palette on

@@ -2,3 +2,3 @@
    <label
-     class="switch"
+     class="switchDarker"
    >
`;

exports[`should change depending on labelText prop 1`] = `
Snapshot Diff:
- label text off
+ label text on

@@ -20,2 +20,7 @@
      </div>
+     <span
+       class="text"
+     >
+       Switch theme
+     </span>
    </label>
`;

exports[`should match snapshot 1`] = `
<DocumentFragment>
  <label
    class="switch"
  >
    <input
      class="checkbox"
      type="checkbox"
    />
    <div
      class="slider"
    >
      <div
        class="round"
        style="background-image: url(sun.svg); opacity: 1;"
      />
      <div
        class="round"
        style="background-image: url(moon.svg); opacity: 0; z-index: 1;"
      />
    </div>
  </label>
</DocumentFragment>
`;

exports[`switching theme should work properly 1`] = `
Snapshot Diff:
- dark theme on
+ light theme on

@@ -12,9 +12,9 @@
        <div
          class="round"
-         style="background-image: url(sun.svg); opacity: 0;"
+         style="background-image: url(sun.svg); opacity: 1;"
        />
        <div
          class="round"
-         style="background-image: url(moon.svg); opacity: 1; z-index: 1;"
+         style="background-image: url(moon.svg); opacity: 0; z-index: 1;"
        />
      </div>
`;
